src/relay/backend/contrib/libtorch/libtorch_codegen.cc

```diff
@@ -0,0 +1,147 @@
+
+bool TorchOpRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
+                const TypeReporter& reporter) {
+  const auto* sfattrs = attrs.as<TorchFunctionAttrs>();
+  std::stringstream str(sfattrs->serialized_function);
+  torch::jit::Module mod = torch::jit::load(str);
+
+  std::vector<torch::jit::IValue> inputs;
+  for (int i = 0; i < num_inputs; i++) {
+    auto* ty = types[i].as<TensorTypeNode>();
+    ICHECK(ty) << "only accept tensors as inputs";
+    std::vector<int64_t> shape;
+    for (const auto& s : ty->shape) {
+      auto* si = s.as<IntImmNode>();
+      if (!si) {
+        return false;
+      }
+      shape.push_back(si->value);
+    }
+    auto torchScalarType = at::toScalarType(ty->dtype);
+
+    inputs.emplace_back(torch::zeros(shape, at::TensorOptions().dtype(torchScalarType)));
+  }
+  auto res = mod.forward(inputs);
+  auto res_t = res.toTensor();
+  ICHECK((int)types.size() == num_inputs + 1) << "only single output supported";
+  Array<PrimExpr> res_sizes;
+  for (int d = 0; d < res_t.dim(); d++) {
+    res_sizes.push_back(IntImm(DataType::Int(32), res_t.size(d)));
+  }
+  reporter->Assign(types[num_inputs], TensorType(res_sizes, DataType(at::getDLDataType(res_t))));
+  return true;
+}
+
+RELAY_REGISTER_OP("torch_op")
+    .set_support_level(99)
+    .add_type_rel("TorchOpRel", TorchOpRel)
+    .set_attrs_type<TorchFunctionAttrs>();
+
+Expr MakeTorchOp(Array<Expr> args, const std::string& serialized_function) {
+  static const Op& op = Op::Get("torch_op");
+  auto attrs = make_object<TorchFunctionAttrs>();
+  attrs->serialized_function = serialized_function;
+  attrs->len = serialized_function.size();
+  return Call(op, args, Attrs(attrs), {});
+}
+
+TVM_REGISTER_GLOBAL("relay.op._make.torchop").set_body_typed(MakeTorchOp);
+
+/*!
+ * \brief The external compiler/codegen tool. It takes a Relay expression/module and
+ * compile it into a runtime module.
+ */
+runtime::Module TorchCompiler(const ObjectRef& ref) {
+  ICHECK(ref->IsInstance<FunctionNode>()) << "The input ref is expected to be a Relay function.";
+  Function func = Downcast<Function>(ref);
+  std::string func_name = backend::GetExtSymbol(func);
+
+  ICHECK(func.defined()) << "Input error: expect a Relay function.";
+  const auto* call = func->body.as<CallNode>();
+  ICHECK(call) << "Expected call node\n";
+  const auto* op_node = call->op.as<OpNode>();
+  ICHECK(op_node) << "Expect OpNode, but got " << call->op->GetTypeKey();
+  const auto op_name = GetRef<Op>(op_node)->name;
+  ICHECK(op_name == "torch_op") << "Unsupported op: " << AsText(call->op, false) << "\n";
+
+  const auto* attrs = call->attrs.as<TorchFunctionAttrs>();
+
+  // TensorRTJSONSerializer serializer(func_name, func);
+  // serializer.serialize();
+  // std::string graph_json = serializer.GetJSON();
+
+  const auto* pf = runtime::Registry::Get("runtime.torch_runtime_create");
+  ICHECK(pf != nullptr) << "Cannot find Torch runtime module create function.";
```

Comment:
For cpp code, it's better to expose this function in a header and directly call it.

Reply:
Thank you! I was unsure whether runtime and codegen cross-links should go through the registry, I'll fix that.